Feature/refactor into sub dirs

redhug1 opened the pull request:
### What

Refactor into sub directories.
Add state transition diagram for when messages are sent to slack.

### How to review

Visual inspection.
Run: make all
If you have a mermain viewer installed in vscode, you can typically view the file in docs directory my opening it in vscode and on macbook typing command+shift+v

### Who can review

Anyone.

redhug1 committed:
move checker and deployment code into their own sub-directories

redhug1 committed:
Update checker mock test

redhug1 committed:
Move NotifySlack func into own package directory

redhug1 committed:
Add mermaid state transition diagram for when slack messages are sent

redhug1 committed:
Remove 'test' from sack message

redhug1 committed:
Slight adjustment to slack message

jessjenkins commented on checker/checker.go:
```diff
@@ -0,0 +1,178 @@
+package checker
+
+//go:generate moq -stub -out checker_mocks_test.go -pkg checker . DeploymentStateGetter SlackNotifier
```
We normally put the mocks in their own subdir. This helps to stop the mock polluting the `checker` package with wrongly exported functions and variables.
```suggestion
//go:generate moq -stub -out mock/checker.go -pkg mock . DeploymentStateGetter SlackNotifier
```

redhug1 replied:
done

jessjenkins commented on checker/checker.go:
```diff
@@ -0,0 +1,178 @@
+// Notify calls slack.NotifySlack.
+func (RealSlackNotifier) Notify(ctx context.Context, cfg *config.Config, result string, state bool) {
+	slack.NotifySlack(ctx, cfg, result, state)
+}
```
Leave the interface in this package (it's defining something that this package needs) but I'd move the type and method to the `slack` package instead. You can then rename them to `SlackNotifier` there. Also I'd probably then rename the interface `Notifier` because it's using a sack one now but it's not inconceivable we could change to a non-slack one later and this would be easy to do then.

redhug1 replied:
done

jessjenkins commented on checker/checker_test.go:
```diff
@@ -1,10 +1,11 @@
-package main
+package checker
```
Can this be renamed to the test package? Might not be able to if it checks private functions but I always try to if I can.
```suggestion
package checker_test
```
NB. same for [deployment/deployment_test.go](https://github.com/ONSdigital/dis-web-mount-check/pull/5/files#diff-eb2d2773e5ea56b06ee655e7e8088f07236d2066882d3bf8da096e37af0a58a3R1) too.

redhug1 replied:
done

jessjenkins commented on deployment/deployment.go:
```diff
@@ -0,0 +1,195 @@
+		nomadClient: nomadClient,
+		endpoint:    cfg.NomadEndpoint,
+		token:       cfg.NomadToken,
+		slackTest:   cfg.SlackTest,
```
Slightly confusing name as it implies something to do with the Slack messaging system. Maybe call it something like `quickTest`?

redhug1 replied:
No, its perfectly named as it does cause messages (assuming tokens, etc are ok) to appear in slack.
Thats its purpose, should one be having any confidence issues with the app reporting changes of state, this allows a wuick test.

jessjenkins replied:
Just because it happens to result in a slack message due to how it is being used within the app, doesn't mean this code's purpose is "Slack" within the context of this package. The slack integration is handled by the importing package, not by this package itself. I disagree it is appropriately named in the context of what this specific package is doing. Additionally other parts of this function outside of this block also trigger slack messages so why is this part specifically 'slack' when they aren't? What is different about this block? Ie. what is it's specific purpose? The name should indicate this.

jessjenkins commented on deployment/deployment.go:
```diff
@@ -0,0 +1,195 @@
+func (d *Deployment) DeploymentState(ctx context.Context, jobID string) (DeploymentState, error) {
+	if d.slackTest {
+		testCount++
+		if testCount <= 60 {
```
Maybe use constants for values in the quick test?

redhug1 replied:
Tied test code in with filter count.

redhug1 committed:
Adjust slack test to relate to number of apps and filter threshold

redhug1 committed:
Update mock test structure

Put genereated code in mock sub-dir
Change package name of checker_test.go to checker_test

redhug1 committed:
Move type and method into slack package and rename vars to suit

jessjenkins commented on go.mod:
Do not use `replace` in committed code. This is only for local development. It bypasses supply-chain injection security checks.

redhug1 replied:
There currently is not an alternative fix to this, and the replace is replacing the supplier hashicorps code with other hashicorps code - so not convinced in this instance about supply-chain injection issues.

jessjenkins replied:
This is a security issue as `replace` disables hash verification of dependencies. There is also an alternative to fix it and I have sent it to you via a DM.

redhug1 replied:
done

redhug1 committed:
Rename var. Update lib version.

redhug1 committed:
Revert go.mod changes to fix CI checks

redhug1 committed:
fix audit with go.mod updates
